presto-memsql/src/test/java/io/prestosql/plugin/memsql/TestMemSqlTypeMapping.java

```diff
@@ -0,0 +1,341 @@
+        throw new SkipException("TODO");
+    }
+
+    @Test
+    public void testDecimalExceedingPrecisionMaxWithSupportedValues()
+    {
+        // TODO determine whether we need decimal_rounding_mode session property
+        throw new SkipException("TODO");
+    }
+
+    @Test
+    public void testPrestoCreatedParameterizedChar()
+    {
+        memSqlCharTypeTest()
+                .execute(getQueryRunner(), prestoCreateAsSelect("memsql_test_parameterized_char"));
+    }
+
+    @Test
+    public void testMemSqlCreatedParameterizedChar()
+    {
+        memSqlCharTypeTest()
+                .execute(getQueryRunner(), memSqlCreateAndInsert("tpch.memsql_test_parameterized_char"));
+    }
+
+    private DataTypeTest memSqlCharTypeTest()
+    {
+        return DataTypeTest.create()
+                .addRoundTrip(charDataType("char", 1), "")
+                .addRoundTrip(charDataType("char", 1), "a")
+                .addRoundTrip(charDataType(1), "")
+                .addRoundTrip(charDataType(1), "a")
+                .addRoundTrip(charDataType(8), "abc")
+                .addRoundTrip(charDataType(8), "12345678")
+                .addRoundTrip(charDataType(255), "a".repeat(255));
+    }
+
+    @Test
+    public void testMemSqlCreatedParameterizedCharUnicode()
+    {
+        DataTypeTest.create()
+                .addRoundTrip(charDataType(1, CHARACTER_SET_UTF8), "\u653b")
+                .addRoundTrip(charDataType(5, CHARACTER_SET_UTF8), "\u653b\u6bbb")
+                .addRoundTrip(charDataType(5, CHARACTER_SET_UTF8), "\u653b\u6bbb\u6a5f\u52d5\u968a")
+                .execute(getQueryRunner(), memSqlCreateAndInsert("tpch.memsql_test_parameterized_varchar"));
+    }
+
+    @Test
+    public void testPrestoCreatedParameterizedVarchar()
+    {
+        DataTypeTest.create()
+                .addRoundTrip(varcharDataType(10), "text_a")
+                .addRoundTrip(varcharDataType(255), "text_b")
+                .addRoundTrip(varcharDataType(256), "text_c")
+                // TODO https://github.com/prestosql/presto/issues/5454
+//                .addRoundTrip(varcharDataType(65535), "text_d")
+//                .addRoundTrip(varcharDataType(16777215), "text_e")
+//                .addRoundTrip(varcharDataType(16777215), "text_f")
+//                .addRoundTrip(varcharDataType(16777216), "text_g")
+//                .addRoundTrip(varcharDataType(VarcharType.MAX_LENGTH), "text_h")
+//                .addRoundTrip(varcharDataType(), "unbounded")
+                .execute(getQueryRunner(), prestoCreateAsSelect("presto_test_parameterized_varchar"));
+    }
+
+    @Test
+    public void testMemSqlCreatedParameterizedVarchar()
+    {
+        DataTypeTest.create()
+                .addRoundTrip(stringDataType("tinytext", createVarcharType(255)), "a")
+                .addRoundTrip(stringDataType("text", createVarcharType(65535)), "b")
+                .addRoundTrip(stringDataType("mediumtext", createVarcharType(16777215)), "c")
```

Comment:
Would be nice to have constants for these "known" types since almost every test uses one of these.
For now we can start with having the constants in this class and can extract to some new/existing class later.

Reply:
you mean 255, 65535, 16777215?
i think these are special to MySQL (lack of true varchar(x) apparently, or our mapping is wrong) and _could be_ special for MemSQL (but turned out not to be)
in my exp, they are not  commonly meaningful for other connectors.